enhancements/testing/build_root_golang_version_from_repo.md

```diff
@@ -0,0 +1,130 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Today, the image in which binaries are built by ci-operator (`build_root_image`) can be
+configured by either specifying an imagestream or by specifying a Dockerfile in the
+repository to build this image from. In the latler case, the Dockerfile is in the basebranch
+of PRs. This means its not possible to do a PR that atomically changes both contents of the
+repository and the `build_root_image`. This proposal proposes a way to enable changing both
+code and the `build_root_image` atomically with one PR.
+
+
+## Motivation
+
+### Goals
+
+* Allow changing both repository contents and `build_root` in a single PR
+
+### Non-Goals
+
+* Change anything about the concept of the `build_root` or how we build images
+  for promotion
+
+## Background
+
+The ci-operator config has a `build_root` section that can either include an imagestreamtag
+or a Dockerfile. If it is an Dockerfile, a build of that Dockerfile from the base branch of the
+PR will be created. In both cases, the resulting image will be tagged into the `pipeline:root`
+imagestreamtag.
+
+The `pipeline:root` imagestreamtag in turn is used to:
+* Build binaries via the `binary_build_command` and tag them into the `pipeline:bin` imagestreamtag
+* Build test binaries via the `test_binary_build_commands` and tag them into the `pipeline:test-bin` imagestreamtag
+* Build rpms via the `rpm_build_commands` and tag them into the `pipeline:rpms` imagestreamtag
+
+All these tags in turn can then be refenced in tests.
+
+Additionally, if the project promotes, there is one Dockerfile per promoted image. It usually duplicates
+the `build_root` and `binary_build_commands` but copies the resulting binaries in an `ocp/$version:base`
+image.
+
+## Proposal
+
+It is proposed to extend the existing `build_root` section in the ci-operator config
+with a new `from_repository` boolean that will indicate of the image should be inferred
+from the repository:
+```
+build_root:
+	from_repository: true
+```
+
+This setting will be mutually exclusive with all other options in the `build_root` section.
+If set, the `ci-operator-prowgen` cli will be changed to clone the repository into the ci
+operator pod. This in turn allows the `ci-operator` to read the image to be used from the
+repository.
+
+The `build_root_image` must not be a raw docker pullspec but an imagestreamtag, because
+we tag it into the `pipeline:root` imagestreamtag and have found that doing an import for
+every job run causes performance issues in the underlying cluster.
+
+TODO: Agree on which of the two options below is better.
+
+There are two options to to get the imagestreamtag from:
+
+### A new configfile in the repository
+
+Very simple, but duplicates with what is in the Dockerfile used for promotion
+
```

Comment:
Please add that – if we choose this one – there is the risk that both Dockerfile and build_root_image divert, probably by accident.

Possible counter measures: (a) build-machinery checking both golang versions (b) have a check on CI side (special job?)

This is really the only fear I have with this method: make sure the same Golang version is used.

Reply:
@tnozicka opinions?

Reply:
Added that

Reply:
I think both options are important. Convention (inferring) with the ability to override with configuration (new upstream config file). For 90% of cases, I believe inference will work well. 

Strawman:
Perform any ci-operator config replacements. Then, for each promotion Dockerfile, based on the state of the Dockerfile(s) in the PR commit create a set:
1. Is the Dockerfile single stage? If yes, it's likely the golang version doesn't matter. Record `default`.
2. Is it Dockerfile multi-stage? If yes, record each stage's ART equivalent golang builder image. If none are present, record `default`.

Different variations in the computed set we might encounter when using inference:
- `{default}`: Use the latest ART equivalent build_root image. 
- `{rhel-8-golang-1.15}`: Use the build_root image based on this golang builder.
- `{default, rhel-7-golang-1.14}`: Weight specific over default and use built_root image based on `rhel-7-golang-1.14`
- `{rhel-7-golang-1.14, rhel-8-golang-1.15}`: Can't reconcile this intelligently. Raise a clear error indicating that the disparity needs to be addressed or a build_root configuration file needs to be included in the PR.

build_root precedence rules:
1. ci-operator/config build_root
2. upstream build_root configuration file
3. inference

If the build_root istag is defined in the ci-operator config or the new configuration file is present in the PR commit under test, the istag will override the inference. However, if the inference logic computes a non-default golang build_root image, if the override is differs, a caution comment will be added to the PR.

Reply:
Err I think this is vastly out of scope. This enhancement is about being able to configure the build_root image explicitly on the repo and have changes to that in a PR being picked up dynamically. Its not about building any kind of heuristic for build_root.

Reply:
I don't think it should be. If manually managing a configuration file upstream is SOP, it makes every PR opened for upstream Dockerfiles to adjust FROM's for ART golang equivalency wrong by default. Do we want engineers to close these PRs and reopen them with a build_root image specification (who drives upstream to do this)? Do we want them to ignore build_root until things break, as it is today? Or does ART implement the heuristic and include it in the PR? 
An enhancement in this area should consider and document the process it anticipates for the organization since the "simple" approach will propagate the need for hundreds of hours of meaningless, error-prone work across the product.

Reply:
> If manually managing a configuration file upstream is SOP, it makes every PR opened for upstream Dockerfiles to adjust FROM's for ART golang equivalency wrong by default.  Do we want engineers to close these PRs and reopen them with a build_root image specification (who drives upstream to do this)?

I have no idea what you are talking about. We do not have any automation that updates build_root.

> An enhancement in this area should consider and document the process it anticipates for the organization since the "simple" approach will propagate the need for hundreds of hours of meaningless, error-prone work across the product.

Interesting take. The only thing this enhancement does is allowing to source the build_root from the repository under test in order to allow for atomic changes of build_root and repository contents. Nothing in this enhancement prevents writing an automation that updates the build_root, it only becomes slightly harder as there are now three instead of two ways on how the build_root could be set up.

The actual task of writing a build_root update automation however is strictly out of scope for this proposal. I'll update it to clarify that.

Reply:
> We do not have any automation that updates build_root.
Right. But we do open PRs for golang builder changes automatically. There is no way for those PRs to be complete (builder images & build_root) atm because build_root is defined in ci-operator config. I'm just trying to point out that, after the enhancement, they will still be incomplete. 
> The actual task of writing a build_root update automation however is strictly out of scope for this proposal. I'll update it to clarify that.
ack

Reply:
> Please add that – if we choose this one – there is the risk that both Dockerfile and build_root_image divert, probably by accident, e.g. we keep unit testing with Golang 1.13, but build with Golang 1.15. That would have been fatal for the etcd case we had some weeks back.

> Possible counter measures: (a) build-machinery checking both golang versions (b) have a check on CI side (special job?)

> This is really the only fear I have with this method: make sure the same Golang version is used.

@sttts It should be mentioned btw that this issue already exists today, the only way in which this enhancement might change things about that is that it might make writing a verification for this a bit harder, as there are now three instead of two ways in which build_root can be defined. While I agree that we should think about how we are not going to make this harder, adding a verification for this isn't in scope for this proposal either. If we happen to get it for free that's nice, but if we don't, the implementation of this proposal won't cover that.

Reply:
Doesn't this - without any heuristic - solve the issue @jupierce ? The PR that automatically updates the `Dockerfile` to update `FROM` can also update the file in the root of the repo that contains the `build_root` config. Why not?

Reply:
@stevekuznetsov It definitely could and I mention it as an approach above (`Or does ART implement the heuristic and include it in the [alignment] PR?`).  My hope was that this enhancement could explicitly dovetail with the alignment PR proposal: https://github.com/openshift/enhancements/pull/484 . Otherwise, we leave this enhancement as a new, raw potential and kick the can to a future enhancement.

Reply:
Yeah I don't see why this enhancement should do anything more than explicitly written. These additional heuristics can be really trivially built on top.